### PR TITLE
ConstantData

### PR DESCRIPTION
**Pull request by FabioLuporini**
Open to debate

### CODE REVIEW

**Comment by FabioLuporini** [comment]
@mlange05 this test is currently failing: ``TestArithmetic.test_indexed_open_loops``
it looks like this is due to some sort of weird effects of using ``eval`` on expressions; if you run the single test, it works like a charm. Otherwise it fails, probably because it's picking some older ``DenseData``, though I'm not sure...

File: devito/interfaces.py [hunk 1]
```diff
@@ -94,18 +94,51 @@ def __init__(self, \*args, \*\*kwargs):
...
+          ------------------                 ------------------
+          |                |                 |                |
+    ScalarFunction  TensorFunction     ConstantData           |
+                                                              |
+                                                ----------------------------
+                                                |             |            |
+                                            DenseData      TimeData  CompositeData
+                                                                           |
+                                                                       PointData
+
```
Comment (navjotk) [inline devito/interfaces.py]: This is very helpful. Thank you for this.

File: devito/dle/backends/basic.py [hunk 2]
```diff
@@ -89,11 +89,10 @@ def _create_elemental_functions(self, state, **kwargs):
...
                 for i in fsymbols:
-                    if i.is_SymbolicFunction:
-                        handle = "(%s*) %s" % (c.dtype_to_ctype(i.dtype), i.name)
-                    else:
-                        handle = "%s_vec" % i.name
-                    args.append((handle, i))
+                    if i.is_TensorFunction:
+                        args.append(("(%s*)%s" % (c.dtype_to_ctype(i.dtype), i.name), i))
+                    elif i.is_SymbolicData:
+                        args.append(("%s_vec" % i.name, i))
```
Comment (navjotk) [inline devito/dle/backends/basic.py]: Isn't this just `ccode`?

Comment (FabioLuporini) [inline devito/dle/backends/basic.py]: good point, I'll try that !

Comment (FabioLuporini) [inline devito/dle/backends/basic.py]: turns out this is the string for the function call, not the function declaration
we could definitely add it along with ``decl`` ``ccode`` etc, but not part of this PR

**Comment by FabioLuporini** [comment]
I suspect the Python3.5 bug here is due to #323

**Comment by FabioLuporini** [comment]
Confirmed: depends on #323

**Comment by FabioLuporini** [comment]
With the latest rebase, this should have become Green

**Comment by FabioLuporini** [comment]
Nope. Still a few minor issues. Will take care tomorrow...

**Comment by mlange05** [comment]
Just had another look through the recent changes, and I'm still happy. Good to go.

**Comment by FabioLuporini** [comment]
Merged.
